compiler/noirc_evaluator/src/ssa/opt/independent_subgraph_detection.rs

```diff
@@ -0,0 +1,383 @@
+        }
+
+        // Merge ValueIds into sets, where each original small set of ValueIds is merged with another set if they intersect
+        context.merge_sets();
+
+        let function_parameters = function.parameters();
+        let variable_parameters_and_return_values: HashSet<ValueId> = function_parameters
+            .iter()
+            .chain(function.returns())
+            .filter(|&x| match function.dfg[*x] {
+                Value::NumericConstant { .. } => false, // Constant values don't connect elements and can be reused in different subgraphs, so we need to avoid them
+                _ => true,
+            })
+            .copied()
+            .collect();
+
+        let mut connected_sets_indices: HashSet<usize> = HashSet::new();
+
+        // Go through each parameter and each set and check if the set contains the parameter
+        // If it's the case, then that set doesn't present an issue
+        for parameter_or_return_value in variable_parameters_and_return_values.iter() {
+            for (set_index, final_set) in context.value_sets.iter().enumerate() {
+                if final_set.contains(parameter_or_return_value) {
+                    connected_sets_indices.insert(set_index);
+                }
+            }
+        }
+
+        // All the other sets of variables are independent
+        let disconnected_sets_indices: Vec<usize> =
+            HashSet::from_iter(0..(context.value_sets.len()))
+                .difference(&connected_sets_indices)
+                .copied()
+                .collect();
+
+        let all_brillig_generated_values: HashSet<ValueId> =
+            context.brillig_return_to_argument_map.keys().copied().collect();
+
+        // Go through each disconnected set
+        for set_index in disconnected_sets_indices.iter() {
+            let current_set = &context.value_sets[*set_index];
+
+            // Find brillig-generated values in the set
+            let intersection: Vec<ValueId> =
+                all_brillig_generated_values.intersection(current_set).cloned().collect();
+            if intersection.is_empty() {
+                // This is probably a test and the values are optimized in a weird way
+                continue;
+            }
+
+            // Go through all brillig outputs in the set
+            for brillig_output_in_set in intersection.iter() {
+                // Get the inputs that correspond to the output
+                let inputs: HashSet<ValueId> = HashSet::from_iter(
+                    context.brillig_return_to_argument_map[brillig_output_in_set].iter().copied(),
+                );
+
+                // Check if any of them are not in the set
+                let unused_inputs: Vec<ValueId> = inputs.difference(current_set).copied().collect();
+
+                // There is a value not in the set, which means that the inputs/outputs of this call have not been properly constrained
+                if !unused_inputs.is_empty() {
```

Comment:
```suggestion
                let unused_inputs = inputs.difference(current_set).next().is_some();

                // There is a value not in the difference set, which means that the inputs/outputs of this call have not been properly constrained
                if unused_inputs {
```
Avoiding the allocation here - no need to iterate over the entire set either

Reply:
Done